packages/server/src/conversations/api.ts

```diff
@@ -1,83 +1,54 @@
-import { Router } from 'express'
-import { Auth } from '../base/auth/auth'
+import { uuid } from '@botpress/messaging-base'
+import { Response } from 'express'
+import { ApiManager } from '../base/api-manager'
+import { ClientApiRequest } from '../base/auth/client'
 import { Schema } from './schema'
 import { ConversationService } from './service'
 
 export class ConversationApi {
-  constructor(private router: Router, private auth: Auth, private conversations: ConversationService) {}
+  constructor(private conversations: ConversationService) {}
 
-  async setup() {
-    this.router.post(
-      '/conversations',
-      this.auth.client.auth(async (req, res) => {
-        const { error } = Schema.Api.Create.validate(req.body)
-        if (error) {
-          return res.status(400).send(error.message)
-        }
-
-        const { userId } = req.body
-        const conversation = await this.conversations.create(req.client!.id, userId)
-
-        res.send(conversation)
-      })
-    )
+  setup(router: ApiManager) {
+    router.post('/conversations', Schema.Api.Create, this.create.bind(this))
+    router.get('/conversations/:id', Schema.Api.Get, this.get.bind(this))
+    router.get('/conversations/user/:userId', Schema.Api.List, this.list.bind(this))
+    router.get('/conversations/user/:userId/recent', Schema.Api.Recent, this.recent.bind(this))
+  }
 
-    this.router.get(
-      '/conversations/:id',
-      this.auth.client.auth(async (req, res) => {
-        const { error } = Schema.Api.Get.validate(req.params)
-        if (error) {
-          return res.status(400).send(error.message)
-        }
+  async create(req: ClientApiRequest, res: Response) {
+    const userId = req.body.userId as uuid
 
-        const { id } = req.params
-        const conversation = await this.conversations.get(id)
+    const conversation = await this.conversations.create(req.client!.id, userId)
+    res.status(201).send(conversation)
+  }
 
-        if (conversation && conversation.clientId !== req.client!.id) {
-          return res.sendStatus(403)
-        } else if (!conversation) {
-          return res.sendStatus(404)
-        }
+  async get(req: ClientApiRequest, res: Response) {
+    const id = req.params.id as uuid
 
-        res.send(conversation)
-      })
-    )
+    const conversation = await this.conversations.get(id)
+    if (!conversation || conversation.clientId !== req.client!.id) {
+      return res.send(undefined)
```

Comment:
We return undefined if `no value` is a correct value to get. 404 implies a client error. Since our client indicates that the get function can return a value or undefined, then undefined should be a value that the server can return. The get function is both a "get properties of that object" and "check if that object exists"

Reply:
Taken from Wikipedia:

| ..., to indicate that the browser was able to communicate with a given server, but the server could not find what was requested. The error may also be used when a server does not wish to disclose whether it has the requested information.

I don't see how this is a client error.

If we were to implement a client in another language, `undefined` might not make sense. I think we should use standard HTTP code as much as possible so we are not faced with surprises in the future. The more we are close to standard the easier it is to connect with any other standard tool.

Reply:
404 is an **error** message. Trying to fetch an item that does not exist is not defined as an error by the api, because we want this route to be useable to check if an item exists. We return undefined as a null value. If we were to adapt the client to any other language we would just use whatever version of `null` that language supports

Reply:
It shouldn't be the API "job" to define what is an error or not. Since we are defining and standardizing a REST (HTTP) API, it makes sense to stay close to the [RFCs](https://www.rfc-editor.org/rfc/rfc7231#section-6.5.4) whose goal is to define how the API should behave. 404 means undefined in HTTP. An HTTP 200 code means that the server found the value, that the request has succeeded. When we return undefined here, the request has not succeeded in finding the requested value.

I understand your point of view that if the request returns nothing, we should do the same thing, but doing so is not the norm. I don't see any disadvantages of returning 404 like all the other API we might use will do.

Furthermore, since we are going to mainly use our clients, they simply have to handle this abstraction of switching 404s by undefined.

It's your call, but I warn you that the more we distance ourselves from the standard, the more we are prone to have unexpected issues in the future.